src/client/client_impl.cc

```diff
@@ -79,6 +82,224 @@ google::protobuf::util::Status ClientImpl::CreateFile(
+  const size_t chunk_block_size(config_manager_->GetFileChunkBlockSize() * 
+                              common::bytesPerMb);
+  // Record the number of bytes that we already read
+  size_t bytes_read(0);
+  // Record the number of bytes to read
+  size_t remain_bytes(nbytes);
+  // A flag to detect EOF
+  bool eof(false);
+  // A variable that keeps track of the start_offset for each chunk. This is
+  // equal to offset % chunk_block_size for the first chunk and 0 for the 
+  // following chunks
+  size_t chunk_start_offset(offset % chunk_block_size);
+
+  // We simply allocate the necessary size for the buffer. There is a 
+  // possibility that we may not read up to nbytes, but it is ok to pass
+  // this optimization opportunity as it may cost more to realloc
+  void* buffer(malloc(nbytes));
+  if (!buffer) {
+    return google::protobuf::util::Status(
+               google::protobuf::util::error::RESOURCE_EXHAUSTED,
+               "Not enough memory: malloc fails");
+  }
+
+  for (size_t chunk_index = offset / chunk_block_size; 
+       remain_bytes > 0 && !eof; chunk_index++) {
+    // Calculate the bytes to be read, which is the min value of the remaining
+    // bytes and the chunk size
+    size_t bytes_to_read(std::min(remain_bytes, chunk_block_size));
+    auto file_chunk_data_or(ReadFileChunk(filename, chunk_index, 
+                                          chunk_start_offset, bytes_to_read)); 
+    // If one of the chunk's read fails, free the buffer and return 
+    // TODO(someone): there is a corner case here if the EOF happens to be
+    // the last byte of a chunk, then if we read the next chunk we ended up
+    // with reading pass EOF. A fix for this is to include a Status in 
+    // ReadFileChunkReply to indicate an EOF
+    if (!file_chunk_data_or.ok()) {
+      free(buffer);
+      return file_chunk_data_or.status();
+    }
+
+    // Concatenate the read data to buffer
+    size_t chunk_bytes_read(file_chunk_data_or.ValueOrDie().bytes_read());
+    const void* chunk_buffer_read(
+        file_chunk_data_or.ValueOrDie().data().c_str());
+    memmove((char*)buffer + bytes_read, chunk_buffer_read, chunk_bytes_read);
+
+    // Update chunk_start_offset,  bytes_read and remain_bytes counts
+    // Starting from the second chunk, chunk_start_offset is zero
+    chunk_start_offset = 0;
+    bytes_read += chunk_bytes_read;
+    remain_bytes -= chunk_bytes_read;
+
+    // Detect EOF if the actual bytes read < bytes-to-read
+    // TODO(someone): maybe we should include a Status in the 
+    // ReadFileChunkReply to indicate an EOF
+    eof = chunk_bytes_read < bytes_to_read;
+  }
+
+  return std::make_pair(bytes_read, buffer);
+}
+
+void ClientImpl::RegisterChunkServerServiceClient(
+    const std::string& server_address) {
+  LOG(INFO) << "Establishing new connection to chunk server:" << server_address;
+  chunk_server_service_client_[server_address] = 
+      new service::ChunkServerServiceGfsClient(grpc::CreateChannel(
```

Comment:
can we use `std::make_shared<ChunkServerServiceGfsClient>(...)` and same for changing variable declarations to `std::shared_ptr`

Reply:
done